hystrix-contrib/hystrix-javanica/hystrix-javanica-core/src/main/java/com/netflix/hystrix/contrib/javanica/command/MetaHolder.java

```diff
@@ -15,96 +15,94 @@
-@Immutable
-public final class MetaHolder {
-
-    private final HystrixCollapser hystrixCollapser;
-    private final HystrixCommand hystrixCommand;
-    private final DefaultProperties defaultProperties;
-
-    private final Method method;
-    private final Method cacheKeyMethod;
-    private final Method ajcMethod;
-    private final Method fallbackMethod;
-    private final Object obj;
-    private final Object proxyObj;
-    private final Object[] args;
-    private final Closure closure;
-    private final String defaultGroupKey;
-    private final String defaultCommandKey;
-    private final String defaultCollapserKey;
-    private final String defaultThreadPoolKey;
-    private final ExecutionType executionType;
-    private final boolean extendedFallback;
-    private final ExecutionType collapserExecutionType;
-    private final ExecutionType fallbackExecutionType;
-    private final boolean fallback;
-    private boolean extendedParentFallback;
-    private final boolean defaultFallback;
-    private final JoinPoint joinPoint;
-    private final boolean observable;
-    private final ObservableExecutionMode observableExecutionMode;
-
-    private static final Function identityFun = new Function<Object, Object>() {
-        @Nullable
-        @Override
-        public Object apply(@Nullable Object input) {
-            return input;
-        }
-    };
-
-    private MetaHolder(Builder builder) {
-        this.hystrixCommand = builder.hystrixCommand;
-        this.method = builder.method;
-        this.cacheKeyMethod = builder.cacheKeyMethod;
-        this.fallbackMethod = builder.fallbackMethod;
-        this.ajcMethod = builder.ajcMethod;
-        this.obj = builder.obj;
-        this.proxyObj = builder.proxyObj;
-        this.args = builder.args;
-        this.closure = builder.closure;
-        this.defaultGroupKey = builder.defaultGroupKey;
-        this.defaultCommandKey = builder.defaultCommandKey;
-        this.defaultThreadPoolKey = builder.defaultThreadPoolKey;
-        this.defaultCollapserKey = builder.defaultCollapserKey;
-        this.defaultProperties = builder.defaultProperties;
-        this.hystrixCollapser = builder.hystrixCollapser;
-        this.executionType = builder.executionType;
-        this.collapserExecutionType = builder.collapserExecutionType;
-        this.fallbackExecutionType = builder.fallbackExecutionType;
-        this.joinPoint = builder.joinPoint;
-        this.extendedFallback = builder.extendedFallback;
-        this.defaultFallback = builder.defaultFallback;
-        this.fallback = builder.fallback;
-        this.extendedParentFallback = builder.extendedParentFallback;
-        this.observable = builder.observable;
-        this.observableExecutionMode = builder.observableExecutionMode;
-    }
-
-    public static Builder builder() {
-        return new Builder();
-    }
+public abstract class MetaHolder<M extends MetaHolder<M,B>,B extends MetaHolder.Builder<M, B>> {
```

Comment:
Yes, basically you can do cast (isn't so bad):
```java
public abstract class AbstractFactory<T, B extends Foo.Builder<B>> {
    public T create(B builder) {
        return (T) builder.build();
    }
}
```
or without cast

```java
public class Foo {

    private int value;

    public Foo(BasicBuilder builder) {
        value = builder.value;
    }

    interface Buildable<T> {
        T build();
    }

    public static abstract class BasicBuilder<B extends BasicBuilder> {
        private int value = 0;

        public BasicBuilder() {
        }

        public B value(int val) {
            value = val;
            return (B) this;
        }

        abstract Object build();
    }

    public static class Builder extends BasicBuilder<Builder> implements Buildable<Foo> {
        public Foo build() {
            return new Foo(this);
        }
    }

}

/***************************************************************************************/

public class Bar extends Foo {

    private int value;

    protected Bar(Builder builder) {
        super(builder);
    }

    public static class Builder extends Foo.BasicBuilder<Builder> implements Buildable<Bar> {
        private int value;

        public Builder barValue(int value) {
            this.value = value;
            return this;
        }

        public Bar build() {
            return new Bar(this);
        }
    }
}

/***************************************************************************************/

public abstract class AbstractFactory<T, B extends Foo.BasicBuilder<B> & Foo.Buildable<T>> {

    public T create(B builder) {
        return builder.build();
    }

}

/***************************************************************************************/

public class BarFactory extends AbstractFactory<Bar, Bar.Builder> {
}

```

But looks complex to justify one cast, isn't ? your call though